File: 2024/C/day_2.c

#include "aoc.h"

bool is_safe(IntArray nums) {
    IntArray deltas = {0};
    deltas.array = (int*)malloc((nums.size-1) * sizeof(int));
    deltas.size = nums.size-1;
    for (size_t i = 1; i < nums.size; i++) {
        deltas.array[i-1] = nums.array[i] - nums.array[i-1];
    }

    if (deltas.array[0] < 0) {
        for (size_t i = 0; i < deltas.size; i++) {
            deltas.array[i] = -deltas.array[i];
        }
    }

    for (size_t i = 0; i < deltas.size; i++) {
        if ((deltas.array[i] > 3) || (deltas.array[i] < 1)) {
            free_int_array(deltas);
            free_int_array(nums);
            return false;
        }
    }
    
    free_int_array(deltas);
    free_int_array(nums);
    return true;
}

int main() {
    char* input = read_input_file("../inputs/day_2.txt");
    StringArray lines = split_string(input, "\n");
    int count = 0;
    for (size_t i = 0; i < 1001; i++) {
        IntArray nums = get_nums(lines.array[i]);
        if (nums.size < 5) {
            free_int_array(nums);
            printf("Skipping line %i\n", i);
            continue;
        }
        if (is_safe(nums)) {
            count++;
        }
    }
    printf("Part 1: %i\n", count);
    free_string_array(lines);
    free(input);
    return 0;
}

File: 2024/rust/src/bin/day_2.rs

fn main() {
    let input: &str = include_str!("../../../inputs/day_2.txt");
    let reports = parse_input(input);
    println!("Part 1: {}", part1(&reports));
    println!("Part 2: {}", part2(&reports));
}

fn parse_input(input: &str) -> Vec<Vec<i32>> {
    let mut reports = Vec::new();
    for line in input.lines() {
        let levels = line.split_whitespace()
            .map(|s| s.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();
        reports.push(levels);
    }
    reports
}

fn valid(levels: &Vec<i32>) -> bool {
    let mut deltas: Vec<i32> = Vec::new();
    for i in 0..levels.len() - 1 {
        let delta = levels[i + 1] - levels[i];
        deltas.push(delta);
    }
    if deltas[0] < 0 {
        deltas = deltas.iter().map(|x| -x).collect();
    }
    deltas.iter().all(|&x| 1 <= x && x <= 3)
}

fn part1(reports: &Vec<Vec<i32>>) -> i32 {
    let mut count = 0;
    for levels in reports {
        if valid(levels) {
            count += 1;
        }
    }
    count
}

fn part2(reports: &Vec<Vec<i32>>) -> i32 {
    let mut count = 0;
    for levels in reports {
        for i in 0..levels.len() {
            let new_levels = {
                let mut new_levels = levels[..i].to_vec();
                new_levels.extend(levels[i + 1..].to_vec());
                new_levels
            };
            if valid(&new_levels) {
                count += 1;
                break;
            }
        }
    }
    count
}
